### DISCUSSION TITLE
Is it possible to declare our own utility types?

### DISCUSSION THREAD

**Discussion by rzvc** [vimeo/psalm#8001] [Plugin development]
I would like to create some utility types that I see Psalm is missing, and according to the discussions in the issue section, won't be implemented anytime soon.

Is it possible to do that via the plugin system? If yes, can someone please point me in the right direction?

Thank you.

**Comment by orklah** [comment]
Hi, it's not possible right now, and I don't want to give hope on that, it won't happen without a rewrite of Psalm.

Every type operation need to be aware of the other existing types. For example, if you create a new special string type, a simple thing like a concatenation need to be aware how your type must behave. Does concatenation on your type must keep your type or must it revert to a string?

Should asserting that your type is not empty change your type into a `non-empty-string` or do you consider that your type must prevail because it's more detailed?

Every single operation, every time a type is included in an union that could contain a parent type should have a special piece of code dedicated to handle your special type. It's a pretty massive work and each interaction need to be thought of in depth

**Reply by rzvc** [reply]
@orklah,

Thanks for the reply.

I don't intend to do anything that low level. What I need is a way of processing types (`type_in` => `type_out` kind of thing), so if I was able to define a function that I can use as a type, that would be amazing.

For example:
```php
$plugin->registerTypeUtility('awesome-type', function(?Union $other) : Union
{
	if (!is_null($other))
		return new Union([ new TString(), new TInt(), ...$other->getAtomicTypes() ]);
	else
		return new Union([ new TString(), new TInt() ]);
});
```
Now I can use it:
```php
class Foo
{
	/** @var awesome-type */
	public $x;

	/** @var awesome-type<bool> */
	public $y;
}
```
This example is trivial, but having this functionality would allow for some really powerful type processing.

Edit: This should be very easy to implement on the Psalm side too, because Psalm, can forget about `awesome-type` as soon as it uses it. All it would care about would be the resulting type.

**Reply by orklah** [reply]
I see, it would be an alias for an union. I guess what you would need to control is here: https://github.com/vimeo/psalm/blob/0bdf9425e3f63a0259e6f28de6cc6396791f6fa7/src/Psalm/Internal/Type/TypeParser.php#L384

This line is the place where we are parsing a type and we end up with a value. we already handled the literals (string, float, int) and we end up calling Atomic::create() because we assume the keyword will be an Atomic. You'll be here with something that isn't an Atomic. So when Atomic::create returns an exception, you'll be able to catch the error and call your callback here

**Reply by rzvc** [reply]
Thank you. That sounds good, I will look into it.

**Reply by AndrolGenhald** [reply]
If we do want to add support for this (and it seems like perhaps it wouldn't be _too_ difficult), we should think through the plugin API for defining these types. I would guess we'd want it to look to the end user like it's basically templates, but should the plugin just receive a variadic argument of `Union`s, or should it have a way to define the expected number and type of templates?

We should also look into providing native utility types [like typescript's](https://www.typescriptlang.org/docs/handbook/utility-types.html), since many of those would probably be useful and should be trivial to implement once a plugin API is implemented.

**Reply by rzvc** [reply]
> I would guess we'd want it to look to the end user like it's basically templates, but should the plugin just receive a variadic argument of `Union`s, or should it have a way to define the expected number and type of templates?

If it covers all cases, then I wouldn't mind Psalm taking care of that, but I can think of several edge cases where that wouldn't be possible so if it was handled on Psalm's side it would surely be limiting. So if I had a vote, I'd vote for the variadic argument and the Plugin would deal with everything.



> We should also look into providing native utility types [like typescript's](https://www.typescriptlang.org/docs/handbook/utility-types.html), since many of those would probably be useful and should be trivial to implement once a plugin API is implemented.

That's pretty much what I'm trying to achieve.

**Reply by AndrolGenhald** [reply]
> If it covers all cases, then I wouldn't mind Psalm taking care of that, but I can think of several edge cases where that wouldn't be possible so if it was handled on Psalm's side it would surely be limiting.

I was thinking the same, perhaps a utility type that takes a variadic number of templates, but on the other hand it would also be good to ensure consistency in the issues emitted when one of the templates isn't of the expected type, or when there are too many or too few templates (I would think we'd want it to match the way templates work with classes). Perhaps good documentation on how to write a plugin adding utility types would be enough.

**Reply by AndrolGenhald** [reply]
One thing to consider is conflicts between utility types and class names. Should utility types be required to include a hyphen? Instead of `Partial<Type>` you'd have `partial-array<array>`, instead of `Omit<Type, "a"|"b">` maybe `omit-keys<array{...}, "a"|"b">` and `omit-properties<object{...}, "a"|"b">`?

**Reply by orklah** [reply]
I made a mistake, Atomic::create will always return a TNamedObject when finding a string it doesn't recognize. This could be then transformed by the callbacks we talked about.

Requiring hyphens would reduce the risk of accidental collisions with user classes, but it would probably make the tool less powerful (we can probably imagine some Plugin that would replace a given class into something else entirely

**Reply by AndrolGenhald** [reply]
> we can probably imagine some Plugin that would replace a given class into something else entirely

idk, that sounds like an anti-feature to me.

**Reply by orklah** [reply]
Possibly yeah, didn't put much thought into it honestly :D

**Reply by rzvc** [reply]
For that an afterNewAnalysisInterface could be useful. There's definitely a use for it - I needed it a couple of days ago.
